core/src/main/java/org/apache/calcite/rel/rules/SubQueryRemoveRule.java

```diff
@@ -915,43 +915,50 @@ private static void matchJoin(SubQueryRemoveRule rule, RelOptRuleCall call) {
     int nFieldsLeft = join.getLeft().getRowType().getFieldCount();
     int nFieldsRight = join.getRight().getRowType().getFieldCount();
 
-    if (inputSet.intersects(ImmutableBitSet.range(0, nFieldsLeft))
-        && inputSet.intersects(ImmutableBitSet.range(nFieldsLeft, nFieldsLeft + nFieldsRight))) {
+
+    ImmutableBitSet inputIntersectsLeftSide = ImmutableBitSet.range(0, nFieldsLeft);
+    ImmutableBitSet inputIntersectsRightSide =
+        ImmutableBitSet.range(nFieldsLeft, nFieldsLeft + nFieldsRight);
+    if (inputSet.intersects(inputIntersectsLeftSide)
+        && inputSet.intersects(inputIntersectsRightSide)) {
+      // The current existential rewrite needs to make join with one side of the origin join and
+      // generate a new condition to replace the on clause. But for RexNode whose operands are
+      // on either side of the join, we can't push them into join. So this rewriting is not
+      // supported.
       return;
     }
 
-    if (inputSet.intersects(ImmutableBitSet.range(0, nFieldsLeft))) {
+    final Set<CorrelationId>  variablesSet = RelOptUtil.getVariablesUsed(e.rel);
+    if (inputSet.intersects(inputIntersectsLeftSide)) {
       builder.push(join.getLeft());
 
-      final Set<CorrelationId>  variablesSet =
-          RelOptUtil.getVariablesUsed(e.rel);
       final RexNode target =
           rule.apply(e, variablesSet, logic, builder, 1, nFieldsLeft, 0);
       final RexShuttle shuttle = new ReplaceSubQueryShuttle(e, target);
 
-      RexNode newCond =
+      final RexNode newCond =
           shuttle.apply(
               RexUtil.shift(join.getCondition(), nFieldsLeft,
                   builder.fields().size() - nFieldsLeft));
-      builder.push(join.getRight())
-          .join(join.getJoinType(), newCond);
+      builder.push(join.getRight());
+      builder.join(join.getJoinType(), newCond);
 
       final int nFields = builder.fields().size();
       ImmutableList<RexNode> fields =
-          builder.fields(ImmutableBitSet.range(0, nFieldsLeft)
```

Comment:
Now I see why you are saving the `ImmutableBitSet`, but I think it's better to save the boolean value of the intersect and rebuild the range here (so basically reverting this part of the change).